pvlib/iotools/bsrn.py

```diff
@@ -0,0 +1,146 @@
+    .. [3] `BSRN Data Retrieval via FTP
+       <https://bsrn.awi.de/data/data-retrieval-via-ftp/>`_
+    .. [4] `BSRN Data Release Guidelines
+       <https://bsrn.awi.de/data/conditions-of-data-release/>`_
+    """
+
+    # Read file and store the starting line number for each logical record (LR)
+    line_no_dict = {}
+    if str(filename).endswith('.gz'):  # check if file is a gzipped (.gz) file
+        with gzip.open(filename, 'rt') as f:
+            for num, line in enumerate(f):
+                if num == 1:  # Get month and year from the 2nd line
+                    start_date = pd.Timestamp(year=int(line[7:11]),
+                                              month=int(line[3:6]), day=1)
+                if line.startswith('*'):  # Find start of all logical records
+                    line_no_dict[line[2:6]] = num  # key is 4 digit LR number
+    else:
+        with open(filename, 'r') as f:
+            for num, line in enumerate(f):
+                if num == 1:  # Get month and year from the 2nd line
+                    start_date = pd.Timestamp(year=int(line[7:11]),
+                                              month=int(line[3:6]), day=1)
+                if line.startswith('*'):  # Find start of all logical records
+                    line_no_dict[line[2:6]] = num
+
+    # Determine start and end line of logical record LR0100 to be parsed
+    start_row = line_no_dict['0100'] + 1  # Start line number
+    # If LR0100 is the last logical record, then read rest of file
+    if start_row-1 == max(line_no_dict.values()):
+        end_row = num  # then parse rest of the file
+    else:  # otherwise parse until the beginning of the next logical record
+        end_row = min([i for i in line_no_dict.values() if i > start_row])
+    nrows = end_row-start_row
+
+    # Read file as a fixed width file (fwf)
+    data = pd.read_fwf(filename, skiprows=start_row, nrows=nrows, header=None,
+                       colspecs=COL_SPECS, na_values=[-999.0, -99.9])
+
+    # Create multi-index and unstack, resulting in one column for each variable
+    data = data.set_index([data.index // 2, data.index % 2])
+    data = data.unstack(level=1).swaplevel(i=0, j=1, axis='columns')
+
+    # Sort columns to match original order and assign column names
+    data = data.reindex(sorted(data.columns), axis='columns')
+    data.columns = BSRN_COLUMNS
+    # Drop empty columns
+    data = data.drop('empty', axis='columns')
+
+    # Change day and minute type to integer
+    data['day'] = data['day'].astype('Int64')
+    data['minute'] = data['minute'].astype('Int64')
+
+    # Set datetime index and localize to UTC
+    data.index = (start_date
+                  + pd.to_timedelta(data['day']-1, unit='d')
+                  + pd.to_timedelta(data['minute'], unit='min'))
+
+    try:
+        data.index = data.index.tz_localize('UTC')  # BSRN timestamps are UTC
+    except TypeError:

```

Comment:
what's the point of this TypeError?

Reply:
also no need for operating on the index: `data = data.tz_localize('UTC')`

Reply:
The try statement was inspired from another pvlib io function. However, with the tz='utc' in the start date as per your suggestion (better solution) this is no longer relevant.